### PR TITLE
1175 Update assignment logic to take level into account

### LINKED ISSUE #1175: Assignment to respect selected level
When level + 1 is assigned and `all levels` filter is on, assigned person is also assigned level 1 assignments.

Sorry could be duplicate, this was raised by @nmadruga:

From telegram:

```
Assignment showing twice once the ProdRegistration is on Stage2 Level 2 (Chefe da secção). It should only be showing the last one
```
Screen Shot 2022-05-04 at 12 34 09 AM

### PR DESCRIPTION
**Pull request by andreievg**
closes #1175 

Created new type `AssignedSectionsByLevel` which is `SectionAssignee` grouped by levels, this type is used in assignment logic.

Noticed a couple of things: 

* Should assigned level 1 be able to assign level above one ?
* Looks like we can approve consolidation when approving just one section (if other sections are not submitted by another reviewer yet)
* Review assignment of level one, that are not assigned, but when level one is fully assigned, are still AVAILABLE ? Is that ok ?

### CODE REVIEW

File: src/utils/types.ts [hunk 1]
```diff
@@ -752,15 +757,15 @@ export type FilterTypeMethod = (filterKey: string, options?: FilterTypeOptions)
 
 export type FilterTypeDefinitions = {
   [filterType in
-    | 'number'
```
Comment (andreievg) [inline src/utils/types.ts:755]: Something is up with my formatter, sorry

File: src/utils/hooks/useUpdateAssignment.tsx [hunk 2]
```diff
@@ -25,23 +26,25 @@ const useUpdateAssignment = ({ fullStructure }: { fullStructure: FullStructure }
...
-              variables: {
-                assignmentId: assignment.id,
-                assignmentPatch,
-              },
-            })
-          )
-        }
-      })
+      assignmentsFiltered
+        .filter(({ level: assignmentLevel }) => assignmentLevel === level)
```
Comment (andreievg) [inline src/utils/hooks/useUpdateAssignment.tsx:30]: only added filter line

File: src/containers/Review/assignment/AssignmentTab.tsx [hunk 3]
```diff
@@ -28,14 +29,8 @@ const AssignmentTab: React.FC<{
     userState: { currentUser },
   } = useUserState()
   const [enableSubmit, setEnableSubmit] = useState<boolean>(false)
-  const [assignedSections, setAssignedSections] = useState<SectionAssignee>(
```
Comment (andreievg) [inline src/containers/Review/assignment/AssignmentTab.tsx:31]: this is relocated to AssignmentRows

File: src/containers/Review/assignment/AssignmentRows.tsx [hunk 4]
```diff
@@ -14,22 +15,30 @@ interface AssignmentRowsProps {
...
-  assignedSections,
-  setAssignedSections,
+  assignedSectionsByLevel,
+  setAssignedSectionsByLevel,
   setEnableSubmit,
   setAssignmentError,
 }) => {
   const { reviewStructuresState } = useReviewStructureState()
 
+  const defaultAssignedSections = Object.values(fullStructure.sections).reduce(
```
Comment (andreievg) [inline src/containers/Review/assignment/AssignmentRows.tsx:34]: Relocated from AssignmentTab

File: src/utils/types.ts (hunk 1 above)
Comment (CarlosNZ) [reply src/utils/types.ts]: I re-auto-formatted that file for you: 0d703ed3

**Comment by CarlosNZ** [comment]
> * Looks like we can approve consolidation when approving just one section (if other sections are not submitted by another reviewer yet)

Yeah, that's pretty bad -- I'd noted it here already: https://github.com/openmsupply/application-manager-web-app/issues/1172#issue-1223446072

> * Should assigned level 1 be able to assign level above one ?

Not sure exactly what you're asking. If you mean, should as assigner be able to assign a higher level than what is currently in progress, then yes -- that's something that Jess was wanting. So a supervisor could go through and do all the assignments at once for the whole stage. Not sure what's actually implemented now though -- if it's possible I think it's more by accident 😊

**Review by CarlosNZ — COMMENTED** [review]
Great, thanks. Looks pretty good I think, but we should probably get @nmadruga to test it out before we merge it, since I'm not 100% clear the conditions it needs to be tested in.

File: src/utils/types.ts (hunk 1 above)
Comment (andreievg) [reply src/utils/types.ts]: thank you

**Comment by nmadruga** [comment]
> * Should assigned level 1 be able to assign level above one ?

Only if this person has `ASSIGN` permissions - which is valid for the whole stage.

**Comment by nmadruga** [comment]
> > * Looks like we can approve consolidation when approving just one section (if other sections are not submitted by another reviewer yet)
> 
> Yeah, that's pretty bad -- I'd noted it here already: [#1172 (comment)](https://github.com/openmsupply/application-manager-web-app/issues/1172#issue-1223446072)

I agree. This shouldn't be happening. Also have a linked issue on our board: https://github.com/openmsupply/application-manager-web-app/issues/1180

**Comment by nmadruga** [comment]
> * Review assignment of level one, that are not assigned, but when level one is fully assigned, are still AVAILABLE ? Is that ok ?

This was changed to simplify the code on re-assignment. Now the AVAILABLE doesn't mean it's possible to assign, but only means it is someone that is assignable for that section - in case no one else is already assigned. Which means, when the assigner clicks on re-assign the system will list all the people AVAILABLE.

**Review by nmadruga — APPROVED** [review]
Thanks! I can see the same code is kept, but fixed and moved the logic to the right places. Which is good - I have tested using the app, so good to me. 

One question: When submitting the reviewAssignments, if you are re-assigning for one level for example, it would still send the updated for for each level. This doesn't seem required, but I don't think it's a problem.

**Pull request merged**